web_monitoring/internetarchive.py

```diff
@@ -381,6 +438,107 @@ def list_versions(self, url, *,
+            -time one. That will always raise a MementoPlaybackError.)
+            Defaults to 43,200 (12 hours).
+
+        Returns
+        -------
+        dict : requests.Response
+            An HTTP response with the content of the memento, including a
+            history of any redirects involved.
+        """
+        with utils.rate_limited(calls_per_second=25, group='get_memento'):
+            # Correctly following redirects is actually pretty complicated. In
+            # the simplest case, a memento is a simple web page, and that's
+            # no problem. However...
+            #   1.  If the response was a >= 400 status, we have to determine
+            #       whether that status is coming from the memento or from the
+            #       the Wayback Machine itself.
+            #   2.  If the response was a 3xx status (a redirect) we have to
+            #       determine the same thing, but it's a little more complex...
+            #       a) If the redirect *is* the memento, its target may be an
+            #          actual memento (see #1) or it may be a redirect (#2).
+            #          The targeted URL is frequently captured anywhere from
+            #          the same second to a few hours later, so it is likely
+            #          the target will result in case 2b (below).
+            #       b) If there is no memento for the requested time, but there
+            #          are mementos for the same URL at another time, Wayback
+            #          *may* redirect to that memento.
+            #          - If this was on the original request, that's *not* ok
+            #            because it means we're getting a different memento
+            #            than we asked for.
+            #          - If the redirect came from a URL that was the target of
+            #            of a memento redirect (2a), then this is expected.
+            #            Before following the redirect, though, we first sanity
+            #            check it to make sure the memento we are redirecting
+            #            to actually came from nearby in time (sometimes
+            #            Wayback will redirect to captures *months* away).
+            history = []
+            urls = set()
+            previous_was_memento = False
+            orginal_url, original_date = memento_url_data(url)
+            response = utils.retryable_request('GET', url, allow_redirects=False, session=self.session)
+            while True:
+                is_memento = 'Memento-Datetime' in response.headers
+
+                if not is_memento:
+                    # If handling a reponse the original memento redirected to,
+                    # a non-memento redirect may be ok. The target URL will
+                    # rarely have been captured at the same time. (See 2b)
+                    playable = False
+                    if previous_was_memento and response.is_redirect:
```

Comment:
Ugh, we can actually have multiple redirects following from a memento and it’s *probably* ok. See this one, for example: http://web.archive.org/web/20180918070531id_/https://www.epa.gov/cleanpowerplan/clean-power-plan-existing-power-plants

```
http://web.archive.org/web/20180918070531id_/https://www.epa.gov/cleanpowerplan/clean-power-plan-existing-power-plants (memento)
http://web.archive.org/web/20180918070531id_/https://www.epa.gov/Energy-Independence (not a memento)
http://web.archive.org/web/20180918073149id_/https://www.epa.gov/energy-independence (not a memento)
http://web.archive.org/web/20180918032922id_/https://www.epa.gov/energy-independence (memento)
```

I’m reasonably certain the `/energy-independence` memento URL from 20180918073149 (3rd in chain) is redirecting because it’s a non-playback-able revisit (it is listed as a revisit in CDX).

So the question is: is a redirect like this acceptable? We don’t allow it for the initial request (in the end is just means the snapshot you asked for couldn’t be played back), but exactly what does this mean when it’s somewhere down the chain of redirects, and you ultimately are sort of looking up a different snapshot? In the strictest sense, we are currently correct in disallowing it — it means the correct capture couldn’t be played back, just like we treat the initial response. It’s iffy whether the URL you get redirected to at this point is likely to still represent the same content, especially if you don’t have a list of CDX results with hashes to guide you (and even then, problematic for redirect mementos.

Reply:
Side note: notice the URL casing changes above. Because everything in wayback is by url_key, which is always lower-case, that should probably be OK. The comparison below on the next line should be case-insensitive.

Reply:
Note: I fixed the case issue, but have not made changes related to the non-playback-that-redirects-to-a-playback-that-is-still-within-redirect_target_window issue.